### PR TITLE
APIServer Distributed Tracing Alpha

### PR DESCRIPTION
**Pull request by dashpole**
Placeholder PR for APIServer tracing.

Enhancement: https://github.com/kubernetes/enhancements/issues/647

### CODE REVIEW

**Comment by PI-Victor** [comment]
this has been removed from the milestone, closing for now, but if there's an exception granted, please re-open it
/close

**Comment by PI-Victor** [comment]
Hello @dashpole :wave: please take a look at [Documenting for a release](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) - PR Ready for Review to get your PR ready for review before Tuesday 20th of July, 11:59 PM PDT. Thank you!

**Comment by dashpole** [comment]
/assign @logicalhan @lilic 

@PI-Victor this is ready for review

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-traces.md:26]:
```suggestion
By default, Kubernetes components export traces using the grpc exporter for OTLP on the [IANA OpenTelemetry port](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=opentelemetry), 4317. As an example, if the collector is running as a sidecar to a Kubernetes component, the following receiver configuration will collect spans and log them to standard output:
```

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-traces.md:52]: nit:
```suggestion
To enable tracing, enable the `APIServerTracing` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) on the kube-apiserver.  Also, provide the kube-apiserver with a tracing configration file with `--tracing-config-file=<path-to-config>`.  This is an example config that records spans for 1 in 10000 requests, and uses the default OpenTelemetry endpoint:
```

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-traces.md:11]: This should use a [feature state shortcode](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#feature-state) to show that it's not a generally available feature.

**Review by sftim — CHANGES_REQUESTED** [review]
Nice to have: document TracingConfiguration in https://kubernetes.io/docs/reference/#config-apis (these are autogenerated)

The two fixes I'd like for sure are:
- fix the spelling of Kubernetes
- add a feature gate shortcode at the top of the page

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (lilic) [inline content/en/docs/concepts/cluster-administration/system-traces.md:14]:
```suggestion
Kubernetes components emit traces in [open telemetry format](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification) using the gRPC exporter and can be collected and routed to tracing backends using the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector#-opentelemetry-collector).
```

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (lilic) [inline content/en/docs/concepts/cluster-administration/system-traces.md:48]:
```suggestion
The kube-apiserver generates spans for incoming HTTP requests, and for outgoing requests to webhooks, etcd, and re-entrant requests.  It propagates the [W3C Trace Context](https://www.w3.org/TR/trace-context/) with outgoing requests but does not make use of the trace context attached to incoming requests, as the kube-apiserver is often a public endpoint.
```

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (lilic) [inline content/en/docs/concepts/cluster-administration/system-traces.md:57]: Since the other option we provide to configure is the endpoint, we should either mention it or just override it here as a way of self-documenting. What do you think?

**Review by lilic — COMMENTED** [review]
Looks good! Couple of comments and suggestions only.

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (sftim) [reply content/en/docs/concepts/cluster-administration/system-traces.md:57]: Could be a commented-out line for the endpoint?

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (sftim) [reply content/en/docs/concepts/cluster-administration/system-traces.md:14]: “OTLP” seems Ok to me.

Comment (lilic) [reply content/en/docs/concepts/cluster-administration/system-traces.md:14]: It's abbreviated from [`OpenTelemetry Protocol`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md)  and I personally prefer using long names in documentation, especially since it's a relatively new concept.

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (lilic) [reply content/en/docs/concepts/cluster-administration/system-traces.md:57]: Yes I think that sounds good as well, just more of a hint that this is the other thing you can override and how to do it.

Comment (dashpole) [reply content/en/docs/concepts/cluster-administration/system-traces.md:60]: Done

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (dashpole) [reply content/en/docs/concepts/cluster-administration/system-traces.md:11]: done

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (dashpole) [reply content/en/docs/concepts/cluster-administration/system-traces.md]: Changed to OpenTelemetry Protocol

**Comment by dashpole** [comment]
Any idea how to generate the reference docs?  I added a placeholder, but still need to fill in the contents.

**Comment by tengqm** [comment]
@dashpole You can file an issue at https://github.com/kubernetes-sigs/reference-docs, and assign it to me. I'll jump onto it later.

**Review by lilic — COMMENTED** [review]
/lgtm

/hold 
In case someone else has any comments.

🎉

File: content/en/docs/reference/config-api/apiserver-tracingconfiguration.v1alpha1.md
Comment (sftim) [inline content/en/docs/reference/config-api/apiserver-tracingconfiguration.v1alpha1.md:6]: This looks wrong - why is the page blank?

I'd expect to have the code in https://github.com/kubernetes-sigs/reference-docs/ generate the content, and have that generated content added as its own commit.

That generation is only a nice-to-have. If it's not yet done, omit `content/en/docs/reference/config-api/apiserver-tracingconfiguration.v1alpha1.md`, don't modify  `content/en/docs/reference/_index.md`, and don't link to that config API reference.

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-traces.md:59]:
```suggestion
# default value
#endpoint: localhost:4317
```

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/system-traces.md:22]: I'd move this to end end of the page.
Also, maybe change:
> There are currently no guarantees of backwards-compatibility for tracing instrumentation.

to
> Until this feature graduates to stable, there are  no guarantees of backwards compatibility
> for tracing instrumentation.

**Review by sftim — CHANGES_REQUESTED** [review]
I'm afraid the TracingConfiguration config API reference isn't correct - it needs to have content.

The other feedback is less important.

**Comment by sftim** [comment]
@dashpole I think you're saying this is ready for review. Could you tweak the PR description so it's definitely clear where things stand?

**Comment by sftim** [comment]
Ah, sorry. That blank page is still there.
/lgtm cancel

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (dashpole) [reply content/en/docs/concepts/cluster-administration/system-traces.md]: Done

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (logicalhan) [inline content/en/docs/concepts/cluster-administration/system-traces.md:16]: s/the/an/ [OpenTelemetry Collector]

**Review by logicalhan — COMMENTED** [review]
/lgtm

**Comment by dashpole** [comment]
Sorry for the delay.  I got part-way through it, and then was sidetracked.  I believe I have addressed all outstanding comments, and I opened an issue for reference docs.

File: content/en/docs/concepts/cluster-administration/system-traces.md
Comment (dashpole) [reply content/en/docs/concepts/cluster-administration/system-traces.md]: done

**Review by logicalhan — COMMENTED** [review]
probably want to squash, but lgtm

**Comment by dashpole** [comment]
squashed

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

**Review by lilic — APPROVED** [review]
/lgtm

🎉

**Comment by PI-Victor** [comment]
thank you!

/approve
the api ref docs are generated in https://github.com/kubernetes/website/pull/29126

**Pull request merged**
